Approved/implicitly-scoped-qubit-allocation.md

```diff
@@ -0,0 +1,342 @@
+
+```qsharp
+operation QubitTuple(n : Int) : Result {
+    using ((a, b, c) = (Qubit(), Qubit[2 * n + 1], Qubit[n])) {
+        // ...
+        return M(q);
+    }
+}
+```
+
+Example 3a: Many qubits allocated with a separate `using` block for each variable.
+
+```qsharp
+operation NestedBlocks(n : Int) : Result {
+    using (a = Qubit()) {
+        using (b = Qubit[2 * n + 1]) {
+            using (c = Qubit[n]) {
+                // ...
+                return M(q);
+            }
+        }
+    }
+}
+```
+
+Example 4a: A clean qubit and a borrowed qubit. This can only be expressed with nested blocks, not with tuple destructuring.
+
+```qsharp
+operation UsingAndBorrowing() : Result {
+    using (a = Qubit()) {
+        borrowing (b = Qubit()) {
+            // ...
+            return M(a);
+        }
+    }
+}
+```
+
+Example 5a: A qubit allocated in a `using` block but released before the end of the outer block.
+
+```qsharp
+operation DifferentLifetime() : Result {
+    mutable r = Zero;
+    using (q = Qubit()) {
+        // Apply operations...
+        set r = M(q);
+    }
+    // Classical post-processing...
+    return r;
+}
+```
+
+## Proposed Modification
+
+The new `use` and `borrow` statements behave identically to `using` and `borrowing`, but the new block is implicitly defined to start at the statement and end at the end of the current block:
+
+```qsharp
+operation FlipCoin() : Result {
+    let x = 10;
+    use q = Qubit();
+    H(q);
+    return M(q);
+}
+```
+
+is equivalent to
+
+```qsharp
+operation FlipCoin() : Unit {
+    let x = 10;
+    using (q = Qubit()) {
+        H(q);
+        return M(q);
+    }
+}
+```
+
+In combination with the new scope statement, the lifetime control provided by `using` can be obtained:
+
+```qsharp
+operation FlipCoin() : Result {
+    mutable r = Zero;
+    {
+        use q = Qubit();
+        H(q);
+        set r = M(q);
+    }
+    return r;
+}
+```
+
+is equivalent to
+
+```qsharp
+operation FlipCoin() : Result {
+    mutable r = Zero;
+    if (true) {
+        using (q = Qubit()) {
+            H(q);
+            set r = M(q);
+        }
+    }
+    return r;
+}
+```
+
+or more simply
+
+```qsharp
+operation FlipCoin() : Result {
+    mutable r = Zero;
+    using (q = Qubit()) {
+        H(q);
+        set r = M(q);
+    }
+    return r;
+}
+```
+
+Each example below is equivalent to the corresponding example in the [Current Status](#current-status) section with the same number.
+
+### Examples
+
+Example 1b: A single qubit allocated and released at the same time as the end of the outer block.
+
+```qsharp
+operation FlipCoin() : Result {
+    use q = Qubit();
+    H(q);
+    return M(q); // q is released here.
+}
+```
+
+Example 2b: Many qubits allocated in a single `use` statement with tuple destructuring.
+
+```qsharp
+operation QubitTuple(n : Int) : Result {
+    use (a, b, c) = (Qubit(), Qubit[2 * n + 1], Qubit[n]);
```

Comment:
Is destructuring required by this proposal, or do existing tuple deconstruction rules apply? In particular, is the following valid?
```Q#
use pair = (Qubit(), Qubit());
PrepareEntangledPair(pair);
```

Reply:
I think that's fine. It's allowed by the current `using` statement and is consistent with `let`.

Reply:
Makes sense, just wanted to ask for clarification. Thanks!